### PR TITLE
Allow the operator to restart the CC when tester processes are unreachable

### PR DESCRIPTION
**Pull request by johscheuer**
# Description

When a cluster contains tester processes and at least one tester process gets unreachable the machine-readable status will contain that unreachable status until the Cluster Controller (CC) is restarted. Until FDB adds better support/handling of those tester processes, we should allow the operator to restart the CC in such cases. If the CC is not restarted the operator is not able to perform some checks as the status is marked as incomplete.

## Type of change

*Please select one of the options below.*

- New feature (non-breaking change which adds functionality)

## Discussion

-

## Testing

Added unit tests and a new e2e test case.

## Documentation

TODO.

## Follow-up

Improve FDB itself to handle those cases better.

### CODE REVIEW

**Comment by johscheuer** [comment]
> ### Result of fdb-kubernetes-operator-pr on Linux CentOS 7
> * Commit ID: [89adf82](https://github.com/FoundationDB/fdb-kubernetes-operator/commit/89adf823245eee19cdf00ae389d723d76d240702)
> * Duration 2:12:57
> * Result: ❌ **FAILED**
> * Error: `Error while executing command: if $fail_test; then exit 1; fi. Reason: exit status 1`
> * [Build Log](https://d1e0l78xbkh3xa.cloudfront.net/fdb-kubernetes-operator/fdb-kubernetes-operator-pr/f5bba052-fe0e-451d-8652-f051f17631c9.gz) terminal output (available for 30 days)
> * [Build Workspace](https://d1e0l78xbkh3xa.cloudfront.net/fdb-kubernetes-operator-pr/f5bba052-fe0e-451d-8652-f051f17631c9/workspace.zip) zip file of the working directory (available for 30 days)

```text
Summarizing 1 Failure:
  [FAIL] Operator Upgrades upgrading a cluster with a random Pod deleted during rolling bounce phase [It] Upgrade from 7.1.53 to 7.3.29 with pods deleted during rolling bounce [e2e, pr]
  /codebuild/output/src3970053497/src/github.com/FoundationDB/fdb-kubernetes-operator/e2e/test_operator_upgrades/operator_upgrades_test.go:155
```

That failure seems unrelated.

**Comment by johscheuer** [comment]
> ### Result of fdb-kubernetes-operator-pr on Linux CentOS 7
> * Commit ID: [89adf82](https://github.com/FoundationDB/fdb-kubernetes-operator/commit/89adf823245eee19cdf00ae389d723d76d240702)
> * Duration 1:48:39
> * Result: ❌ **FAILED**
> * Error: `Error while executing command: if $fail_test; then exit 1; fi. Reason: exit status 1`
> * [Build Log](https://d1e0l78xbkh3xa.cloudfront.net/fdb-kubernetes-operator/fdb-kubernetes-operator-pr/e9952b13-8459-440b-9628-3a84a5bc6d93.gz) terminal output (available for 30 days)
> * [Build Workspace](https://d1e0l78xbkh3xa.cloudfront.net/fdb-kubernetes-operator-pr/e9952b13-8459-440b-9628-3a84a5bc6d93/workspace.zip) zip file of the working directory (available for 30 days)

```text
Summarizing 1 Failure:
  [FAIL] Operator Upgrades upgrading a cluster with a crash looping sidecar process [It] Upgrade from 7.1.53 to 7.3.29 with crash-looping sidecar [e2e, pr]
  /codebuild/output/src1423644072/src/github.com/FoundationDB/fdb-kubernetes-operator/e2e/fixtures/fdb_cluster.go:1323
```

I take a closer look.

File: api/v1beta2/foundationdb_status.go [hunk 1]
```diff
@@ -54,8 +54,18 @@ type FoundationDBStatusMessage struct {
...
 }
 
+// FoundationDBUnreachableProcess provides information about an unreachable process
+type FoundationDBUnreachableProcess struct {
+	// Address provides the address of the unreachable process
+	Address string `json:"address,omitempty"`
+}
+
+//  "address" : "100.82.115.41:4500:tls"
+
```
Comment (johscheuer) [inline api/v1beta2/foundationdb_status.go:68]:
```suggestion

```

File: controllers/bounce_processes.go [hunk 2]
```diff
@@ -326,3 +336,76 @@ func getAddressesForUpgrade(logger logr.Logger, r *FoundationDBClusterReconciler
...
 }
+
+// checkIfClusterControllerNeedsRestart checks if the cluster controller must be restarted. There are a few cases where
+// this might be required. One case is when at least on tester process is running in the cluster and that tester process
+// fails. Currently this leads to the cluster controller reporting unreachable processes and the status incomplete message.
+// Having those messages in the cluster's machine-readable status could block some operations of the operator and the
+// solution to that is to restart the cluster controller process.
+func checkIfClusterControllerNeedsRestart(status *fdbv1beta2.FoundationDBStatus) *fdbv1beta2.ProcessAddress {
+	// If the status contains no cluster messages we can skip further check.
+	if len(status.Cluster.Messages) == 0 {
```
Comment (johscheuer) [inline controllers/bounce_processes.go:347]: We should add a special minimum uptime here for this behaviour to limit the CC restarts. This should be configurable over a command line flag on the operator with the default value of 1h.

File: e2e/test_operator/operator_test.go [hunk 3]
```diff
@@ -1747,4 +1747,67 @@ var _ = Describe("Operator", Label("e2e", "pr"), func() {
...
+			// We will be restarting the CC, so we can ignore this check.
+			availabilityCheck = false
+			spec := fdbCluster.GetCluster().Spec.DeepCopy()
+			spec.ProcessCounts.Test = 1
+			fmt.Println("Adding 1 tester process to the cluster")
+			fdbCluster.UpdateClusterSpecWithSpec(spec)
+			Expect(fdbCluster.WaitForReconciliation()).NotTo(HaveOccurred())
+		})
+
+		It("when the tester process fails", func() {
```
Comment (johscheuer) [inline e2e/test_operator/operator_test.go:1762]: We should add a test case here for the case where the tester is able to connect to the CC but the CC is not able to connect to the CC to better understand the behaviour in this scenario.

File: e2e/test_operator/operator_test.go [hunk 4]
```diff
@@ -1747,4 +1747,150 @@ var _ = Describe("Operator", Label("e2e", "pr"), func() {
 			Expect(cluster.Status.DatabaseConfiguration.RoleCounts.Proxies).NotTo(BeZero())
 		})
 	})
+
+	FWhen("running with tester processes", func() {
```
Comment (brownleej) [inline e2e/test_operator/operator_test.go:1751]: Do we need to clear the focus before merging?

**Review by brownleej — CHANGES_REQUESTED** [review]
This mostly looks good, but I think there's a small change needed in one of the tests.

File: e2e/test_operator/operator_test.go (hunk 4 above)
Comment (johscheuer) [reply e2e/test_operator/operator_test.go:1751]: Thanks for catching that 👍

**Comment by johscheuer** [comment]
> ### Result of fdb-kubernetes-operator-pr on Linux CentOS 7
> * Commit ID: [a2d8d56](https://github.com/FoundationDB/fdb-kubernetes-operator/commit/a2d8d5672b35b919baf855bfda5b8d959f996c4b)
> * Duration 1:50:09
> * Result: ❌ **FAILED**
> * Error: `Error while executing command: if $fail_test; then exit 1; fi. Reason: exit status 1`
> * [Build Log](https://d1e0l78xbkh3xa.cloudfront.net/fdb-kubernetes-operator/fdb-kubernetes-operator-pr/9187891a-cd90-4429-823b-c5cd19b369b1.gz) terminal output (available for 30 days)
> * [Build Workspace](https://d1e0l78xbkh3xa.cloudfront.net/fdb-kubernetes-operator-pr/9187891a-cd90-4429-823b-c5cd19b369b1/workspace.zip) zip file of the working directory (available for 30 days)

```text

Summarizing 1 Failure:
  [FAIL] Operator Upgrades one process is marked for removal [It] Upgrade from 7.1.53 to 7.3.29 [e2e, pr]
  /codebuild/output/src827113698/src/github.com/FoundationDB/fdb-kubernetes-operator/e2e/fixtures/fdb_cluster.go:1323
```

Seems unrelated.

**Review by brownleej — APPROVED** [review]
👍

**Pull request merged**